### PR TITLE
WELD-2743 Fix how Weld creates child CreationalContext for invocations of BM#getReference. Add a test explaining this scenario

### PR DESCRIPTION
**Pull request by manovotn**
JIRA - https://issues.redhat.com/browse/WELD-2743

### CODE REVIEW

**Review by mkouba — APPROVED** [review]
Looks good.

File: impl/src/main/java/org/jboss/weld/manager/BeanManagerImpl.java [hunk 1]
```diff
@@ -663,7 +663,11 @@ private Context internalGetContext(Class<? extends Annotation> scopeType) {
     }
 
     public Object getReference(Bean<?> bean, Type requestedType, CreationalContext<?> creationalContext, boolean noProxy) {
-        if (creationalContext instanceof CreationalContextImpl<?>) {
+        return getReference(bean, requestedType, creationalContext, noProxy, true);
+    }
+
+    public Object getReference(Bean<?> bean, Type requestedType, CreationalContext<?> creationalContext, boolean noProxy, boolean createChildCc) {
```
[the remark this answers is not in the document]
Comment (manovotn) [reply impl/src/main/java/org/jboss/weld/manager/BeanManagerImpl.java]: True, changed to `private`

**Comment by manovotn** [comment]
The fact that the CI is failing is correct because of this fix for Weld will not pass current state of the TCK test (`DependentContextTest#testDependentScopedInterceptorsAreDependentObjectsOfBean`) which is being challenged in https://github.com/jakartaee/cdi-tck/issues/454

We're going to need a new TCK release with either the suggested TCK PR fix or an exlusion to be able to include this fix on Weld side.

**Comment by manovotn** [comment]
The two failing tests are both from our TS (`ClientConversationContextTest` and `Weld1178Test`) and the only fail with latest WFLY used in CI.
If I take latest released WFLY, patch in newest Weld and run then, it passes just fine.
I'm going to have to take a look at what are those tests doing and at latest WFLY changes...

**Comment by manovotn** [comment]
So, `Weld1178Test` turns out to be a blast from the past - still using some `.jsf` files and I suppose WFLY finally dropped support for those. Turning those into `.xhtml` pages makes the test work again.
I am still trying to figure out `ClientConversationContextTest` as that one doesn't seem outdated and no longer returns error code on forcibly passing in non-existend `cid`...

**Comment by manovotn** [comment]
In the end I managed to find a way to rewrite the failing test - it now uses servlet filter to capture the thrown exception and directly write it into the response, allowing us to verify it instead of just expecting a return code. This should pass in latest nightly WFLY as well.

**Pull request merged**